[GitHub Actions] - Verify Release Build Workflow

dannymartinm opened the pull request:
As part of the [Beam Release Process](https://beam.apache.org/contribute/release-guide/), the following PR includes the workflow to run the [verify_release_build.sh](https://github.com/apache/beam/blob/master/release/src/main/scripts/verify_release_build.sh) script described in [step 6. Verify Release Branch - Run automation script](https://beam.apache.org/contribute/release-guide/#run-automation-script-verify_release_buildsh).

Files: 
* Composite action to install GitHub CLI in linux: `.github/actions/setup-gh-cli-linux/action.yml`
* Verify Release Workflow: `.github/workflows/verify_release_build.yml`
* Script to comment Trigger Phrases in batch in the given Pull Request: `scripts/ci/release/comment_pr_trigger_phrases.sh` 
    * `scripts/ci/release/mass_comment.txt` file where are included the same Trigger Phrases that are in the [mass_comment.py](https://github.com/apache/beam/blob/master/release/src/main/scripts/mass_comment.py) file.
    * `scripts/ci/release/jenkins_trigger_phrases.txt` where are included the Release Gradle Build job and some PostCommit and PreCommits jobs that will not be migrated to GitHub Actions in the first iteration of the massive migration. 
    
`* Note`: Currently, the `verify_release_build` workflow is pointing to the `mass_comment.txt` file when commenting Trigger Phrases as we don't have enabled Trigger Phrases for PostCommit and PreCommit GitHub Actions workflows yet and also, we haven't migrated those jobs. Once there are in place, feel free to update the workflow to call the PostCommits and PreCommits workflows as well as pointing to the `jenkins_trigger_phrases.txt` file which includes some of those jobs that are missing in GitHub Actions. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

dannymartinm commented:
Hi @Abacn and @damccorm , hope all is good.
Could you please help with a code review?
Thank you in advance!

Abacn commented:
@dannymartinm Thanks. Looking

Abacn commented on CI.md:
```diff
@@ -125,6 +125,13 @@ Service Account shall have following permissions ([IAM roles](https://cloud.goog
+#### Verify Release Build - [verify_release_build.yml](.github/workflows/verify_release_build.yml)
+| Job                          | Description                                                                                   | Pull Request Run | Direct Push/Merge Run | Scheduled Run | Requires GCP Credentials |
+|------------------------------|-----------------------------------------------------------------------------------------------|------------------|-----------------------|---------------|--------------------------|
+| Verify Release Build         | Verifies full life cycle of Gradle Build and all PostCommit/PreCommit tests against Release Branch on Jenkins.                   | No               | No                    | No            | No                       |
```
Nit: "on Jenkins" documentation outdated

dannymartinm replied:
Could you please add more details on what should I do in this line? Thank you!

damccorm replied:
```suggestion
| Verify Release Build         | Verifies full life cycle of Gradle Build and all PostCommit/PreCommit tests against Release Branch on CI.                   | No               | No                    | No            | No                       |
```

Could we just make it generic?

Abacn commented:
Missing Apache 2.0 license for these files:
```
17:16:56 > Task :rat
17:16:56 Files with unapproved licenses:
17:16:56  - /home/jenkins/jenkins-slave/workspace/beam_PreCommit_RAT_Commit/src/scripts/ci/release/jenkins_trigger_phrases.txt
17:16:56  - /home/jenkins/jenkins-slave/workspace/beam_PreCommit_RAT_Commit/src/scripts/ci/release/mass_comment.txt
```
I understand these files are just trigger phrases. Could organize it into this folder orginization `"**/test/resources/**/*.txt"` (or simply embed them into script) as it will be excluded from license check: https://github.com/apache/beam/blob/610ccba65b51c522fa29c79a5d4b85dce73c2b8c/build.gradle.kts#L53.

dannymartinm commented:
> Missing Apache 2.0 license for these files:
> 
> ```
> 17:16:56 > Task :rat
> 17:16:56 Files with unapproved licenses:
> 17:16:56  - /home/jenkins/jenkins-slave/workspace/beam_PreCommit_RAT_Commit/src/scripts/ci/release/jenkins_trigger_phrases.txt
> 17:16:56  - /home/jenkins/jenkins-slave/workspace/beam_PreCommit_RAT_Commit/src/scripts/ci/release/mass_comment.txt
> ```
> 
> I understand these files are just trigger phrases. Could organize it into this folder orginization `"**/test/resources/**/*.txt"` (or simply embed them into script) as it will be excluded from license check:
> 
> https://github.com/apache/beam/blob/610ccba65b51c522fa29c79a5d4b85dce73c2b8c/build.gradle.kts#L53
> 
> .

Updated and moved txt files into /test/resources directory. Thanks!

Abacn reviewed (approved):
LGTM

dannymartinm commented on .github/workflows/verify_release_build.yml:
```diff
@@ -0,0 +1,78 @@
+        description: Flag to comment the Jenkins Trigger Phrases in the new PR
+        required: true
+env:
+  WORKING_BRANCH: postcommit_validation_pr # Is it always the same branch name? or should we need to add it as input?
```
I was thinking on that if every time the Release process finishes, the PRs and branches created for it must be deleted by the committer. In that case, we won't have branch conflicts.

damccorm replied:
I'm ok with that, but we should do some validation to make sure that the branch doesn't exist when we start the workflow. If this step gets missed then you could get weird behavior where the validation is run on an old version.

Also, we should remove this comment

dannymartinm commented on scripts/ci/release/comment_pr_trigger_phrases.sh:
```diff
@@ -0,0 +1,26 @@
+GITHUB_PR=$1
+while IFS= read -r trigger_phrase
+do
+    gh pr comment "$GITHUB_PR" --body "$trigger_phrase"
```
Totally, that's why I was thinking on having for the moment the `jenkins_trigger_phrases.txt` which includes the trigger phrases of the Jenkins Jobs that won't be migrated to GitHub Actions in this first iteration/migration. 

On the other hand, if you think commenting is out of scope and it would be better to not have it yet, I can remove the `Comment Jenkins Trigger Phrases in PR` step of the workflow.

Thank you!

damccorm replied:
I think leaving it in is the right thing. My point is just that once we've fully migrated to GHA, AFAIK we won't have trigger phrases available to us and that functionality will need to be replaced in this process.

dannymartinm replied:
Yes, I agree. I think it would be great to start thinking of some GHA triggers out of the Continuous Integration ones. 
Thanks Danny for your comments!

damccorm reviewed (approved):
Thanks!
